source/common/stats/tag_extractor_impl.cc

```diff
@@ -169,5 +180,110 @@ bool TagExtractorRe2Impl::extractTag(absl::string_view stat_name, std::vector<Ta
   return false;
 }
 
+TagExtractorTokensImpl::TagExtractorTokensImpl(absl::string_view name, absl::string_view tokens,
+                                               absl::string_view substr)
+    : TagExtractorImplBase(name, tokens, substr), tokens_(absl::StrSplit(tokens, '.')),
+      match_index_(findMatchIndex(tokens_)) {
+  prefix_ = tokens_[0];
+  substr_index_ = 0;
+
+  for (uint32_t i = 1; i < tokens_.size(); ++i) {
+    absl::string_view token = tokens_[i];
+    if (token == "**") {
+      // Can't infer a token index if we see a pattern that can have a variable number of dots.
+      break;
+    }
+    if (substr_.empty() && token != "*" && token != "$") {
+      substr_ = token;
+    }
+    if (substr_ == token) {
+      substr_index_ = i;
+      break;
+    }
+  }
+}
+
+uint32_t TagExtractorTokensImpl::findMatchIndex(const std::vector<std::string>& tokens) {
+  for (uint32_t i = 0; i < tokens.size(); ++i) {
+    if (tokens[i] == "$") {
+      return i;
+    }
+  }
+  return 0;
+}
+
+bool TagExtractorTokensImpl::extractTag(TagExtractionContext& context, std::vector<Tag>& tags,
+                                        IntervalSet<size_t>& remove_characters) const {
+  PERF_OPERATION(perf);
+
+  const std::vector<absl::string_view>& tokens = context.tokens();
+  if ((substr_index_ == 0 && substrMismatch(context.name())) ||
+      (substr_index_ != 0 &&
+       (substr_index_ >= tokens.size() || tokens[substr_index_] != substr_))) {
+    PERF_RECORD(perf, "tokens-skip", name_);
+    PERF_TAG_INC(skipped_);
+    return false;
+  }
```

Comment:
BTW, is it still beneficial to do a substring match first?

Reply:
I think there is but I'm going to re-measure to quantify that.

Reply:
Well it seems like it's a wash; it helped some cases to do the pre-check and hurt others, so I just killed the precheck.

Reply:
hi  you tagged the wrong person 😅